test/gtest/trace.hpp

```diff
@@ -0,0 +1,252 @@
+        {{1024, 1024}, true},                 // cont large case
+        {{1, 10}, true},                      // cont small case
+        {{34, 20}, true},                     // cont small case
+    };
+    // clang-format on
+}
+
+template <typename T>
+struct TraceFwdTest : public ::testing::TestWithParam<TraceTestCase>
+{
+protected:
+    void SetUp() override
+    {
+        auto&& handle   = get_handle();
+        trace_config    = GetParam();
+        auto gen_value1 = [](auto...) { return prng::gen_descreet_uniform_sign<T>(1e-2, 100); };
+        auto gen_value2 = [](auto...) { return prng::gen_descreet_uniform_sign<T>(1e-2, 99); };
+
+        auto in_dims    = trace_config.input_dim;
+        auto in_strides = trace_config.ComputeStrides();
+        input           = tensor<T>{in_dims, in_strides}.generate(gen_value1);
+
+        auto out_lengths = std::vector<size_t>{1};
+
+        output = tensor<T>{out_lengths};
+        std::fill(output.begin(), output.end(), std::numeric_limits<T>::quiet_NaN());
+
+        ref_output = tensor<T>{out_lengths};
+        std::fill(ref_output.begin(), ref_output.end(), std::numeric_limits<T>::quiet_NaN());
+
+        ws_sizeInBytes =
+            miopen::trace::GetTraceForwardWorkspaceSize(handle, input.desc, output.desc);
+        if(ws_sizeInBytes == static_cast<size_t>(-1))
+            GTEST_SKIP();
+
+        if(ws_sizeInBytes != 0)
+        {
+            std::vector<size_t> workspace_dims;
+            workspace_dims.push_back(ws_sizeInBytes / sizeof(float));
+
+            workspace = tensor<float>{workspace_dims};
+            std::fill(workspace.begin(), workspace.end(), static_cast<float>(0));
+
+            workspace_dev = handle.Write(workspace.data);
+        }
+
+        input_dev  = handle.Write(input.data);
+        output_dev = handle.Write(output.data);
+    }
+
+    void RunTest()
+    {
+        auto&& handle = get_handle();
+
+        miopenStatus_t status;
+
+        cpu_trace_forward<T>(input, ref_output);
+        status = miopen::trace::TraceForward(handle,
+                                             workspace_dev.get(),
+                                             ws_sizeInBytes,
+                                             input.desc,
+                                             input_dev.get(),
+                                             output.desc,
+                                             output_dev.get());
+
+        EXPECT_EQ(status, miopenStatusSuccess);
+
+        output.data = handle.Read<T>(output_dev, output.data.size());
+    }
+
+    double GetTolerance()
+    {
+        double tolerance = std::numeric_limits<T>::epsilon() * 10;
+        return tolerance;
+    }
+
+    void Verify()
+    {
+        double threshold = GetTolerance();
+        auto error       = miopen::rms_range(ref_output, output);
+
+        EXPECT_TRUE(error < threshold * 10) << "Error output beyond tolerance Error: " << error
```

Comment:
ASSERT_EQ(miopen::range_distance(ref_output), miopen::range_distance(output));
EXPECT_LT

Reply:
I have inserted `EXPECT_EQ(miopen::range_distance(ref_output), miopen::range_distance(output));`
I confuse whether exist any differences between `EXPECT_LT` and `EXPECT_TRUE` in this case?

Reply:
EXPECT_LT use EXPECT_PRED* that was recommended to use: "Use EXPECT_PRED* instead of EXPECT_TRUE if possible". https://github.com/ROCm/MIOpen/wiki/GTest-development

Reply:
updated.